### PR TITLE
darwin packages: Get ready for cross

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

If things build fine with `stdenvNoCC`, let them use that. If tools
might be prefixed, prepare for that.

This is a mass rebuild, but I don't think the final stdenv should be any different in contents, so I think it's safe.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/darwin/apple-source-releases/Csu/default.nix [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 
-appleDerivation {
   prePatch = ''
     substituteInPlace Makefile \
       --replace /usr/lib /lib \
       --replace /usr/local/lib /lib \
       --replace /usr/bin "" \
       --replace /bin/ "" \
-      --replace "CC = " "CC = cc #" \
+      --replace "CC = " "CC = ${stdenv.cc.targetPrefix or ""}cc #" \
```
Comment (veprbl) [inline pkgs/os-specific/darwin/apple-source-releases/Csu/default.nix:11]:
```suggestion
      --replace "CC = " "#" \
```

File: pkgs/os-specific/darwin/apple-source-releases/configd/default.nix [hunk 2]
```diff
@@ -127,78 +127,78 @@ appleDerivation {
...
-    cc -I. -Ihelper -Iderived -F. -c VPNPrivate.c -o VPNPrivate.o
-    cc -I. -Ihelper -Iderived -F. -c VPNConfiguration.c -o VPNConfiguration.o
-    cc -I. -Ihelper -Iderived -F. -c VPNTunnel.c -o VPNTunnel.o
-    cc -I. -Ihelper -Iderived -F. -c derived/helperUser.c -o helperUser.o
-    cc -I. -Ihelper -Iderived -F. -c reachability/SCNetworkReachabilityServer_client.c -o SCNetworkReachabilityServer_client.o
-    cc -I. -Ihelper -Iderived -F. -c reachability/rb.c -o rb.o
-    cc -I. -Ihelper -Iderived -F. -c derived/SystemConfiguration_vers.c -o SystemConfiguration_vers.o
-
-    cc -dynamiclib *.o -install_name $out/Library/Frameworks/SystemConfiguration.framework/SystemConfiguration -dead_strip -framework CoreFoundation -single_module -o SystemConfiguration.framework/SystemConfiguration
+    ${stdenv.cc.targetPrefix or ""}cc -I. -Ihelper -Iderived -F. -c SCSchemaDefinitions.c -o SCSchemaDefinitions.o
```
Comment (veprbl) [inline pkgs/os-specific/darwin/apple-source-releases/configd/default.nix:130]:
```suggestion
    $CC -I. -Ihelper -Iderived -F. -c SCSchemaDefinitions.c -o SCSchemaDefinitions.o
```

File: pkgs/os-specific/darwin/apple-source-releases/default.nix [hunk 3]
```diff
@@ -220,17 +220,22 @@ let
   macosPackages_11_0_1 = import ./macos-11.0.1.nix { inherit applePackage'; };
   developerToolsPackages_11_3_1 = import ./developer-tools-11.3.1.nix { inherit applePackage'; };
 
-  applePackage' = namePath: version: sdkName: sha256:
+  applePackage'_ = stdenv: namePath: version: sdkName: sha256:
```
Comment (veprbl) [inline pkgs/os-specific/darwin/apple-source-releases/default.nix:223]: Probably don't need to introduce a function that we only call once

File: pkgs/os-specific/darwin/apple-source-releases/default.nix [hunk 4]
```diff
@@ -220,17 +220,22 @@ let
...
     let
       pname = builtins.head (lib.splitString "/" namePath);
-      appleDerivation = appleDerivation' pname version sdkName sha256;
-      callPackage = self.newScope { inherit appleDerivation; };
+      appleDerivation' = stdenv: appleDerivation'' stdenv pname version sdkName sha256;
+      appleDerivation = appleDerivation' stdenv;
+      callPackage = self.newScope { inherit appleDerivation' appleDerivation; };
```
Comment (veprbl) [inline pkgs/os-specific/darwin/apple-source-releases/default.nix:228]: How about `appleDerivationNoCC`?

File: pkgs/os-specific/darwin/apple-source-releases/libresolv/default.nix [hunk 5]
```diff
@@ -1,37 +1,39 @@
...
-    cc -I. -c res_init.c
-    cc -I. -c res_mkquery.c
-    cc -I. -c res_mkupdate.c
-    cc -I. -c res_query.c
-    cc -I. -c res_send.c
-    cc -I. -c res_sendsigned.c
-    cc -I. -c res_update.c
-    cc -dynamiclib -install_name $out/lib/libresolv.9.dylib -current_version 1.0.0 -compatibility_version 1.0.0 -o libresolv.9.dylib *.o
+  buildPhase = lib.optionalString (!headersOnly) ''
+    ${stdenv.cc.targetPrefix or ""}cc -I. -c dns_util.c
```
Comment (veprbl) [inline pkgs/os-specific/darwin/apple-source-releases/libresolv/default.nix:9]:
```suggestion
    $CC -I. -c dns_util.c
```

File: pkgs/os-specific/darwin/apple-source-releases/default.nix (hunk 4 above)
Comment (Ericson2314) [reply pkgs/os-specific/darwin/apple-source-releases/default.nix:228]: Well `appleDerivation'` takes an arbitrary `stdenv`. I think I would like that to just replace `appleDerivation` eventually.

File: pkgs/os-specific/darwin/apple-source-releases/default.nix (hunk 3 above)
Comment (Ericson2314) [reply pkgs/os-specific/darwin/apple-source-releases/default.nix]: True.

**Comment by Ericson2314** [comment]
@Mic92 No, but I believe it is "getting the mass rebuilds out of the way". I have another commit with an actual attempt at that which I think might work better after @thefloweringash's darwin -> darwin attempt.

**Comment by Ericson2314** [comment]
Rebuilt darwin stdenv and hello just to double check.

**Pull request merged**